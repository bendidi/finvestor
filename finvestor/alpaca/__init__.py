from finvestor.alpaca.api import AlpacaClient

__all__ = "AlpacaClient"
