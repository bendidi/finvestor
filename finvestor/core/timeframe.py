import re
import typing as tp
from datetime import timedelta

import attr
from pydantic import ValidationError
from pydantic.datetime_parse import parse_duration as parse_duration_pydantic
from pydantic.errors import DurationError
from pydantic.fields import ModelField

_NUM_DAYS_IN_A_MONTH = 30.417
_NUM_DAYS_IN_A_YEAR = 365
_UNITS_REGEX = r"(?P<val>\d+(\.\d+)?)(?P<unit>(mo|s|m|h|d|w|y)?)"
_UNITS = {
    "s": "seconds",
    "m": "minutes",
    "h": "hours",
    "d": "days",
    "w": "weeks",
}


def parse_duration(value: tp.Any) -> timedelta:
    try:
        return parse_duration_pydantic(value)
    except DurationError:
        pass

    if value.lower() == "mtd":
        return timedelta(days=_NUM_DAYS_IN_A_MONTH)
    if value.lower() == "ytd":
        return timedelta(days=_NUM_DAYS_IN_A_YEAR)

    kwargs: tp.Dict[str, float] = {}
    for match in re.finditer(_UNITS_REGEX, value, flags=re.I):
        unit = match.group("unit").lower()
        val = float(match.group("val"))
        if unit == "mo":
            unit = "d"
            val = val * _NUM_DAYS_IN_A_MONTH
        elif unit == "y":
            unit = "d"
            val = val * _NUM_DAYS_IN_A_YEAR
        if unit not in _UNITS:
            raise DurationError()
        if _UNITS[unit] in kwargs:
            kwargs[_UNITS[unit]] += val
        else:
            kwargs[_UNITS[unit]] = val

    if not kwargs:
        raise DurationError()
    return timedelta(**kwargs)  # type: ignore


TimeFrameType = tp.TypeVar("TimeFrameType")


@attr.s
class TimeFrame(tp.Generic[TimeFrameType]):

    duration = attr.ib(eq=parse_duration, order=parse_duration)

    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, value: tp.Any, field: ModelField):
        if not field.sub_fields:
            if not isinstance(value, cls):
                raise TypeError(
                    f"No subField provided, Expected <class '{cls.__name__}'>, "
                    f"got {type(value)} (value={value})"
                )
            return value

        sub_field = field.sub_fields[0]

        if not isinstance(value, cls):
            value = cls(value)

        _, error = sub_field.validate(value.duration, {}, loc="duration")
        if error:
            raise ValidationError([error], cls)  # type: ignore
        return value
