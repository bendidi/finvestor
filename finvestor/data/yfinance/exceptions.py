class EmptyResponse(Exception):
    pass


class ErrorResponse(Exception):
    pass
