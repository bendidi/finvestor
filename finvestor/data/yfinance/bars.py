import asyncio
import logging
from datetime import datetime
from typing import Any, Dict, List, Optional, Tuple

import numpy as np
import pytz
from httpx import AsyncClient
from pydantic import validate_arguments

from finvestor.core import config
from finvestor.data.yfinance import YF_HEADERS
from finvestor.data.yfinance.exceptions import EmptyResponse, ErrorResponse
from finvestor.data.yfinance.utils import (
    ValidInterval,
    ValidRange,
    coerce_end_time_to_interval,
    get_valid_intervals_for_datetime,
    get_valid_intervals_for_range,
    time_symbol_to_timedelta,
)
from finvestor.models import Bar, Bars

logger = logging.getLogger(__name__)


async def get_quotes(
    ticker: str,
    *,
    client: AsyncClient,
    start: Optional[datetime],
    end: Optional[datetime],
    range: Optional[ValidRange],
    interval: ValidInterval = "1m",
    include_prepost: bool = False,
    coerce_interval: bool = False,
) -> Tuple[List[float], Dict]:

    params: Dict[str, Any] = {}

    if start is not None:
        assert start.tzinfo == pytz.utc, (
            f"Only timezone aware datetimes with tz={pytz.UTC} are supported, "
            f"got: start<{start.tzinfo}>"
        )
        if end is not None:
            assert end.tzinfo == pytz.utc, (
                f"Only timezone aware datetimes with tz={pytz.UTC} are supported, "
                f"got: end<{end.tzinfo}>"
            )
            assert end > start, (
                f"Provided end datetime ({end}) should be larger than start "
                f"datetime ({start})"
            )
            if coerce_interval:
                end = coerce_end_time_to_interval(ticker, interval, start, end)
        else:
            end = start + time_symbol_to_timedelta(interval)
        params["period1"] = int(datetime.timestamp(start))
        params["period2"] = int(datetime.timestamp(end))
    else:
        params["range"] = range

    params["interval"] = interval
    params["includePrePost"] = include_prepost
    params["events"] = "div,splits"
    resp = await client.get(
        url=f"{config.YF_BASE_URL}/v8/finance/chart/{ticker}",
        params=params,
        headers=YF_HEADERS,
    )
    response = resp.json()

    error = response.get("chart", {}).get("error")
    if error:
        logger.error(f"Error response from Y-Finance [{ticker}]: {error}")
        raise ErrorResponse(error)

    result = response.get("chart", {}).get("result", [])
    if not len(result):
        logger.error(
            f"Empty response from Y-Finance [{ticker}]: range={range} | "
            f"interval={interval} | start={start} | end={end}"
        )
        raise EmptyResponse(f"Empty response: {response}")

    timestamps = result[0].get("timestamp")
    if not timestamps:
        logger.error(
            f"Empty quotes from Y-Finance [{ticker}]: range={range} | "
            f"interval={interval} | start={start} | end={end}"
        )
        raise EmptyResponse(f"Empty quotes: {response}")

    ohlc = result[0]["indicators"]["quote"][0]
    return timestamps, ohlc


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_bars(
    ticker: str,
    *,
    client: AsyncClient,
    start: Optional[datetime] = None,
    end: Optional[datetime] = None,
    range: Optional[ValidRange] = None,
    interval: ValidInterval = "1m",
    include_prepost: bool = False,
    coerce_interval: bool = True,
) -> Bars:

    assert (
        range is not None or start is not None
    ), "please provide either a range or start[-end] datetimes"

    if range is not None and coerce_interval:
        _valid_intervals = get_valid_intervals_for_range(interval, range)
    elif start is not None and coerce_interval:
        _valid_intervals = get_valid_intervals_for_datetime(interval, start)
    else:
        _valid_intervals = [interval]

    if interval not in _valid_intervals:
        logger.debug(
            f"[{ticker}]: {interval} interval is not supported by yahoo-finance "
            f"for provided range ({range}), or start/end datetimes ({start}/{end}). "
            f"Supported intervals for your request: {_valid_intervals}"
        )
    errors = []
    for _interval in _valid_intervals:
        try:
            timestamps, ohlc = await get_quotes(
                ticker,
                client=client,
                start=start,
                end=end,
                range=range,
                interval=_interval,
                include_prepost=include_prepost,
                coerce_interval=coerce_interval,
            )
            break
        except EmptyResponse as error:
            errors.append(error)
            continue
    else:
        raise EmptyResponse(errors)
    bars = Bars(
        __root__=[
            Bar(
                timestamp=timestamp,
                close=close if close is not None else np.nan,
                high=high if high is not None else np.nan,
                low=low if low is not None else np.nan,
                open=open if open is not None else np.nan,
                volume=volume if volume is not None else np.nan,
                interval=_interval,
            )
            for timestamp, volume, open, close, low, high in zip(
                timestamps,
                ohlc["volume"],
                ohlc["open"],
                ohlc["close"],
                ohlc["low"],
                ohlc["high"],
            )
        ]
    )
    return bars


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_latest_bar(
    ticker: str,
    *,
    client: AsyncClient,
    interval: ValidInterval = "1m",
    include_prepost: bool = True,
) -> Bar:

    if interval.endswith("m") or interval.endswith("h"):
        range = "1d"
    elif interval.endswith("d"):
        range = "1mo"
    else:
        range = "6mo"
    bars = await get_bars(
        ticker,
        client=client,
        include_prepost=include_prepost,
        range=range,  # type: ignore
    )
    return bars[-1]


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_bar_at_timestamp(
    ticker: str,
    *,
    client: AsyncClient,
    timestamp: datetime,
) -> Bar:
    assert timestamp.tzinfo == pytz.utc, (
        f"Only timezone aware datetimes with tz={pytz.UTC} are supported, "
        f"got: timestamp<{timestamp.tzinfo}>"
    )
    interval: ValidInterval = "1m"
    start = timestamp.replace(second=0)

    bars = await get_bars(ticker, client=client, start=start, interval=interval)
    return bars[0]


async def get_closest_price_at_timestamp(
    ticker: str,
    *,
    client: AsyncClient,
    timestamp: datetime,
) -> float:
    bar = await get_bar_at_timestamp(ticker, client=client, timestamp=timestamp)
    bar_open_date = bar.timestamp
    bar_close_date = bar_open_date + time_symbol_to_timedelta(bar.interval)
    open_delta = timestamp - bar_open_date
    close_delta = bar_close_date - timestamp
    if open_delta < close_delta:
        return bar.open
    return bar.close


if __name__ == "__main__":
    from finvestor.core import setup_logging
    from finvestor.etoro.utils import parse_etoro_datetime

    setup_logging()

    async def main():
        ticker = "FB"
        async with AsyncClient() as client:
            date = parse_etoro_datetime("28/09/2021 15:13:03")
            bars = await get_bar_at_timestamp(ticker, client=client, timestamp=date)
            print(bars)

    asyncio.run(main())
