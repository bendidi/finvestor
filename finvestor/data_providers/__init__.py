from finvestor.data_providers.yahoo_finance.assets import get_asset, get_isin
from finvestor.data_providers.schemas import *
from finvestor.data_providers.bars import *

__all__ = (
    "get_asset",
    "get_isin",
    "Bars",
    "Bar",
    "Asset",
    "get_alpaca_bars",
    "get_yahoo_finance_bars",
    "get_bars",
    "get_latest_bar",
    "get_bar_at_timestamp",
    "get_price_at_timestamp",
)
