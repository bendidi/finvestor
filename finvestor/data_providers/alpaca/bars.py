import logging
import typing as tp
from datetime import datetime, timedelta
from urllib.parse import urljoin

import pytz
from httpx import AsyncClient
from pydantic import root_validator, validate_arguments, validator

from finvestor import config
from finvestor.core.timeframe import TimeFrame
from finvestor.data_providers.alpaca.auth import auth
from finvestor.data_providers.exceptions import EmptyBars
from finvestor.data_providers.schemas import Bar, Bars, BarsRequestParams
from finvestor.data_providers.utils import (
    get_closest_value_to_timedelta,
    parse_duration,
)

logger = logging.getLogger(__name__)


ApcaValidIntervals = tp.Literal[
    "1Min", "2Min", "5Min", "15Min", "30Min", "1Hour", "3Hour", "1Day"
]
ApcaTickerTypes = tp.Literal["stock", "crypto"]
APCA_VALID_INTERVALS: tp.Tuple[ApcaValidIntervals, ...] = tp.get_args(
    ApcaValidIntervals
)
APCA_INTERVAL_MAP: tp.Mapping[timedelta, ApcaValidIntervals] = {
    parse_duration(interval): interval for interval in APCA_VALID_INTERVALS
}
APCA_TICKER_TYPE_TO_URLPATH: tp.Dict[ApcaTickerTypes, str] = {
    "crypto": "/v1beta1/crypto/{ticker}/bars",
    "stock": "/v2/stocks/{ticker}/bars",
}


class AlpacaBarsRequestParams(BarsRequestParams):
    interval: TimeFrame[ApcaValidIntervals] = TimeFrame("1Min")
    historical_data_delay: timedelta = timedelta(0)
    limit: int = 10000
    adjustment: tp.Literal["raw", "split", "dividend", "all"] = "all"

    @validator("interval", pre=True)
    def convert_interval(cls, interval: TimeFrame) -> TimeFrame:
        if not isinstance(interval, TimeFrame):
            interval = TimeFrame(interval)
        if interval.duration in APCA_VALID_INTERVALS:
            return interval
        duration = get_closest_value_to_timedelta(
            APCA_INTERVAL_MAP, parse_duration(interval.duration)
        )
        return TimeFrame(duration)

    @root_validator
    def convert_period_to_start_end(cls, values):
        start = values.get("start")
        end = values.get("end")
        valid_delay: timedelta = values["historical_data_delay"]
        now = pytz.utc.localize(datetime.utcnow())
        if start is not None and end is not None:
            delta_delay = now - end
            if valid_delay.total_seconds() > 0 and delta_delay <= valid_delay:
                raise ValueError(
                    "Provided 'end' value does not respect ALPACA historical data delay"
                    f", provided end={end} has a diff of {delta_delay} < {valid_delay}"
                )
            return values

        # TODO: doesn't take into account days were the market is not Open
        # (holidays, weekends, ..)
        # maybe have a look at: https://github.com/rsheftel/pandas_market_calendars
        period: TimeFrame = values.get("period")
        if period is not None:
            end = now - valid_delay
            start = end - parse_duration(period.duration)
            return {**values, "start": start, "end": end}
        return values

    def __call__(
        self,
        ticker_type: ApcaTickerTypes = "stock",
    ) -> tp.Dict[str, tp.Union[str, int]]:
        assert self.start is not None
        assert self.end is not None
        assert ticker_type in tp.get_args(ApcaTickerTypes)
        params = {
            "start": self.start.isoformat(),
            "end": self.end.isoformat(),
            "limit": self.limit,
            "timeframe": self.interval.duration,
        }
        if ticker_type == "stock":
            params["adjustment"] = self.adjustment

        return params


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_alpaca_bars(
    ticker: str,
    request_params: tp.Union[BarsRequestParams, tp.Dict[str, tp.Any]],
    *,
    historical_data_delay: timedelta = config.APCA_HISTORICAL_DATA_DELAY,
    client: AsyncClient,
    ticker_type: ApcaTickerTypes = "stock",
) -> Bars:

    if isinstance(request_params, BarsRequestParams):
        request_params = request_params.dict()

    alpaca_request_params = AlpacaBarsRequestParams(
        **request_params, historical_data_delay=historical_data_delay
    )
    alpaca_request_params_dict = alpaca_request_params(ticker_type=ticker_type)
    logger.debug(
        f"[ALPACA] Getting data for ticker '{ticker}' ({ticker_type}) "
        f"with params: {alpaca_request_params_dict}"
    )

    urlpath = APCA_TICKER_TYPE_TO_URLPATH[ticker_type].format(ticker=ticker)

    url = urljoin(config.APCA_API_DATA_URL, urlpath)

    bars: tp.List[tp.Dict[str, tp.Any]] = []
    next_page_token = ""
    while next_page_token is not None:
        if next_page_token:
            alpaca_request_params_dict["page_token"] = next_page_token
        resp = await client.get(url, auth=auth, params=alpaca_request_params_dict)
        if resp.status_code != 200:
            logger.error(
                f"Alpaca (ticker={ticker}) responded with code <{resp.status_code}>: "
                f"{resp.json()}"
            )
            resp.raise_for_status()
        response = resp.json()
        next_page_token = response.get("next_page_token")
        resp_bars = response.get("bars", [])
        if not bars and not resp_bars:
            raise EmptyBars(ticker, response)
        if resp_bars is not None:
            bars.extend(resp_bars)

    return Bars(
        __root__=[
            Bar(
                timestamp=_bar["t"],
                close=_bar["c"],
                high=_bar["h"],
                low=_bar["l"],
                open=_bar["o"],
                volume=_bar["v"],
                interval=alpaca_request_params.interval.duration,
            )
            for _bar in bars
        ]
    )
