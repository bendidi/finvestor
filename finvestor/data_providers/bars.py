from datetime import datetime
from typing import Optional, Tuple

import pytz
from httpx import AsyncClient
from pydantic import validate_arguments

from finvestor.core import config
from finvestor.data_providers.alpaca.bars import (
    get_alpaca_bars,
    is_before_apca_hstorical_delay,
)
from finvestor.data_providers.schemas import Bar, Bars
from finvestor.data_providers.utils import (
    ValidInterval,
    ValidPeriod,
    get_valid_period_for_interval,
    time_period_to_timedelta,
)
from finvestor.data_providers.yahoo_finance.bars import get_yahoo_finance_bars


@validate_arguments(config=dict(arbitrary_types_allowed=True))
async def get_bars(
    ticker: str,
    *,
    client: AsyncClient,
    start: Optional[datetime] = None,
    end: Optional[datetime] = None,
    period: Optional[ValidPeriod] = None,
    interval: ValidInterval = "1m",
) -> Bars:
    if config.DATA_PROVIDER == "alpaca":
        return await get_alpaca_bars(
            ticker,
            client=client,
            start=start,
            end=end,
            period=period,
            interval=interval,
        )
    return await get_yahoo_finance_bars(
        ticker,
        client=client,
        start=start,
        end=end,
        period=period,
        interval=interval,
    )


async def get_latest_bar(
    ticker: str,
    *,
    client: AsyncClient,
    interval: ValidInterval = "1m",
) -> Bar:
    if (
        config.DATA_PROVIDER == "alpaca"
        and config.APCA_HISTORICAL_DATA_DELAY_SECONDS.total_seconds() == 0
    ):
        bars = await get_alpaca_bars(
            ticker,
            client=client,
            interval=interval,
            period=get_valid_period_for_interval(interval),
        )
    else:
        bars = await get_yahoo_finance_bars(
            ticker,
            client=client,
            interval=interval,
            period=get_valid_period_for_interval(interval),
        )
    return bars[-1]


async def get_bar_at_timestamp(
    ticker: str,
    *,
    client: AsyncClient,
    timestamp: datetime,
    interval: ValidInterval = "1m",
) -> Bar:
    assert timestamp.tzinfo == pytz.utc, (
        f"Only timezone aware datetimes with tz={pytz.UTC} are supported, "
        f"got: timestamp<{timestamp.tzinfo}>"
    )
    start = timestamp.replace(second=0)
    end = start + time_period_to_timedelta(interval)
    if is_before_apca_hstorical_delay(end):
        bars = await get_bars(
            ticker, client=client, start=start, end=end, interval=interval
        )
    else:
        bars = await get_yahoo_finance_bars(
            ticker, client=client, start=start, end=end, interval=interval
        )
    return bars[0]


async def get_price_at_timestamp(
    ticker: str,
    *,
    client: AsyncClient,
    timestamp: datetime,
) -> Tuple[datetime, float]:
    bar = await get_bar_at_timestamp(
        ticker, client=client, timestamp=timestamp, interval="1m"
    )
    bar_open_date = bar.timestamp
    bar_close_date = bar_open_date + time_period_to_timedelta(bar.interval)
    open_delta = timestamp - bar_open_date
    close_delta = bar_close_date - timestamp
    if open_delta < close_delta:
        return bar_open_date, bar.open
    return bar_close_date, bar.close
