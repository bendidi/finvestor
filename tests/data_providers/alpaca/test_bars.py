from contextlib import nullcontext as does_not_raise
from datetime import datetime, timedelta

import pytest
import pytz

from finvestor.data_providers.alpaca.bars import (
    AlpacaBarsRequestParams,
    get_alpaca_bars,
)
from finvestor.data_providers.exceptions import EmptyBars
from finvestor.data_providers.timeframe import TimeFrame


def test_AlpacaBarsRequestParams_data_validation(alpaca_bars_request_params_test_cases):
    (
        interval,
        period,
        start,
        end,
        kwargs,
        expecation,
    ) = alpaca_bars_request_params_test_cases
    with expecation:
        AlpacaBarsRequestParams(
            interval=interval, period=period, start=start, end=end, **kwargs
        )


@pytest.mark.parametrize(
    "period,historical_data_delay",
    [
        ("1d 10m", timedelta(minutes=15)),
        ("1Min", timedelta(minutes=1)),
        ("25Min", timedelta(minutes=100)),
    ],
)
def test_AlpacaBarsRequestParams_convert_period_to_start_end(
    period, historical_data_delay
):

    params = AlpacaBarsRequestParams(
        period=period, historical_data_delay=historical_data_delay
    )

    now = datetime.now(tz=pytz.UTC)
    assert params.period.duration == period
    assert TimeFrame(params.end - params.start) == params.period
    # not that accurate of a test, but good enough
    assert now - params.end > historical_data_delay


@pytest.mark.parametrize(
    "interval,expected",
    [
        ("1Min", "1Min"),
        ("2m", "2Min"),
        ("25Min", "30Min"),
        ("1d", "1Day"),
        (timedelta(days=1, hours=2), "1Day"),
    ],
)
def test_AlpacaBarsRequestParams_convert_interval(interval, expected):

    period = "5d"
    params = AlpacaBarsRequestParams(period=period, interval=interval)
    assert params.interval.duration == expected


@pytest.mark.parametrize(
    "ticker_type,kwargs",
    [("stock", {"adjustment": "all"}), ("crypto", {})],
)
def test_AlpacaBarsRequestParams___call___(ticker_type, kwargs):
    start = datetime.now(tz=pytz.UTC) - timedelta(days=6)
    end = start + timedelta(days=5)
    req_params = AlpacaBarsRequestParams(start=start, end=end, interval="1Day")

    assert req_params(ticker_type=ticker_type) == {
        "start": start.isoformat(),
        "end": end.isoformat(),
        "limit": 10000,
        "timeframe": "1Day",
        **kwargs,
    }


@pytest.mark.parametrize(
    "ticker,start,end,ticker_type,limit,expectation",
    [
        pytest.param(
            "TSLA",
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc),
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc)
            + timedelta(minutes=5),
            "stock",
            10000,
            does_not_raise(),
            id="TSLA-stock-2021-11-26T15-00-00",
        ),
        pytest.param(
            "TSLA",
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc),
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc)
            + timedelta(minutes=5),
            "stock",
            2,
            does_not_raise(),
            id="TSLA-stock-2021-11-26T15-00-00-limit-2",
        ),
        pytest.param(
            "BTCUSD",
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc),
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc)
            + timedelta(minutes=5),
            "crypto",
            10000,
            does_not_raise(),
            id="BTCUSD-crypto-2021-11-26T15-00-00",
        ),
        pytest.param(
            "ADAUSD",  # ADA not supported at times of writing on alpaca
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc),
            datetime(year=2021, month=11, day=26, hour=15, tzinfo=pytz.utc)
            + timedelta(minutes=5),
            "crypto",
            10000,
            pytest.raises(EmptyBars),
            id="ADAUSD-crypto-2021-11-26T15-00-00",
        ),
        pytest.param(
            "TSLA",  # market is closed on a sunday so empty bar
            datetime(year=2021, month=11, day=28, hour=15, tzinfo=pytz.utc),
            datetime(year=2021, month=11, day=28, hour=15, tzinfo=pytz.utc)
            + timedelta(minutes=5),
            "stock",
            10000,
            pytest.raises(EmptyBars),
            id="TSLA-stock-2021-11-28T15-00-00",
        ),
    ],
)
@pytest.mark.vcr
async def test_get_alpaca_bars(
    snapshot, async_client, ticker, start, end, ticker_type, limit, expectation
):
    request_params = {"start": start, "end": end, "interval": "1Min", "limit": limit}
    with expectation:
        bars = await get_alpaca_bars(
            ticker,
            request_params,
            historical_data_delay=timedelta(0),
            client=async_client,
            ticker_type=ticker_type,
        )
        assert bars == snapshot
