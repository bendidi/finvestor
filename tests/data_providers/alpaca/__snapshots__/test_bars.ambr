# name: test_get_alpaca_bars[BTCUSD-crypto-2021-11-26T15-00-00]
  <class 'dict'> {
    '__root__': <class 'list'> [
      <class 'dict'> {
        'close': 54643.57,
        'high': 54725.49,
        'interval': TimeFrame(duration='1Min'),
        'low': 54612.0,
        'open': 54725.48,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 0, tzinfo=datetime.timezone.utc),
        'volume': 35.45622815,
      },
      <class 'dict'> {
        'close': 54631.5,
        'high': 54631.5,
        'interval': TimeFrame(duration='1Min'),
        'low': 54631.5,
        'open': 54631.5,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 0, tzinfo=datetime.timezone.utc),
        'volume': 0.0092,
      },
      <class 'dict'> {
        'close': 54583.85,
        'high': 54640.9,
        'interval': TimeFrame(duration='1Min'),
        'low': 54583.08,
        'open': 54640.9,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 1, tzinfo=datetime.timezone.utc),
        'volume': 46.55203187,
      },
      <class 'dict'> {
        'close': 54565.3,
        'high': 54587.6,
        'interval': TimeFrame(duration='1Min'),
        'low': 54565.3,
        'open': 54587.6,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 1, tzinfo=datetime.timezone.utc),
        'volume': 5.1228,
      },
      <class 'dict'> {
        'close': 54595.0,
        'high': 54595.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 54559.0,
        'open': 54595.0,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 1, tzinfo=datetime.timezone.utc),
        'volume': 7.0208,
      },
      <class 'dict'> {
        'close': 54580.01,
        'high': 54618.39,
        'interval': TimeFrame(duration='1Min'),
        'low': 54577.01,
        'open': 54583.84,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 2, tzinfo=datetime.timezone.utc),
        'volume': 16.5809722,
      },
      <class 'dict'> {
        'close': 54569.3,
        'high': 54572.5,
        'interval': TimeFrame(duration='1Min'),
        'low': 54553.5,
        'open': 54567.6,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 2, tzinfo=datetime.timezone.utc),
        'volume': 9.1074,
      },
      <class 'dict'> {
        'close': 54607.0,
        'high': 54607.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 54591.0,
        'open': 54591.0,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 2, tzinfo=datetime.timezone.utc),
        'volume': 0.2749,
      },
      <class 'dict'> {
        'close': 54543.98,
        'high': 54607.91,
        'interval': TimeFrame(duration='1Min'),
        'low': 54543.98,
        'open': 54580.01,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 3, tzinfo=datetime.timezone.utc),
        'volume': 22.08330112,
      },
      <class 'dict'> {
        'close': 54522.6,
        'high': 54558.6,
        'interval': TimeFrame(duration='1Min'),
        'low': 54522.6,
        'open': 54558.6,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 3, tzinfo=datetime.timezone.utc),
        'volume': 2.7594,
      },
      <class 'dict'> {
        'close': 54564.0,
        'high': 54564.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 54564.0,
        'open': 54564.0,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 3, tzinfo=datetime.timezone.utc),
        'volume': 0.0021,
      },
      <class 'dict'> {
        'close': 54550.0,
        'high': 54574.14,
        'interval': TimeFrame(duration='1Min'),
        'low': 54535.65,
        'open': 54544.0,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 4, tzinfo=datetime.timezone.utc),
        'volume': 32.78464713,
      },
      <class 'dict'> {
        'close': 54527.6,
        'high': 54527.6,
        'interval': TimeFrame(duration='1Min'),
        'low': 54509.7,
        'open': 54509.7,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 4, tzinfo=datetime.timezone.utc),
        'volume': 2.0092,
      },
      <class 'dict'> {
        'close': 54552.0,
        'high': 54563.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 54552.0,
        'open': 54563.0,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 4, tzinfo=datetime.timezone.utc),
        'volume': 0.5511,
      },
      <class 'dict'> {
        'close': 54537.76,
        'high': 54580.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 54500.0,
        'open': 54548.99,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 5, tzinfo=datetime.timezone.utc),
        'volume': 59.71624911,
      },
      <class 'dict'> {
        'close': 54493.4,
        'high': 54542.3,
        'interval': TimeFrame(duration='1Min'),
        'low': 54466.3,
        'open': 54529.8,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 5, tzinfo=datetime.timezone.utc),
        'volume': 8.237,
      },
      <class 'dict'> {
        'close': 54561.0,
        'high': 54561.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 54561.0,
        'open': 54561.0,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 5, tzinfo=datetime.timezone.utc),
        'volume': 0.02,
      },
    ],
  }
---
# name: test_get_alpaca_bars[TSLA-stock-2021-11-26T15-00-00-limit-2]
  <class 'dict'> {
    '__root__': <class 'list'> [
      <class 'dict'> {
        'close': 1093.3,
        'high': 1095.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 1091.64,
        'open': 1093.62,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 0, tzinfo=datetime.timezone.utc),
        'volume': 88087.0,
      },
      <class 'dict'> {
        'close': 1090.66,
        'high': 1093.6826,
        'interval': TimeFrame(duration='1Min'),
        'low': 1088.91,
        'open': 1093.04,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 1, tzinfo=datetime.timezone.utc),
        'volume': 53702.0,
      },
      <class 'dict'> {
        'close': 1087.21,
        'high': 1090.6399,
        'interval': TimeFrame(duration='1Min'),
        'low': 1087.2,
        'open': 1090.2,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 2, tzinfo=datetime.timezone.utc),
        'volume': 50956.0,
      },
      <class 'dict'> {
        'close': 1089.9759,
        'high': 1090.49,
        'interval': TimeFrame(duration='1Min'),
        'low': 1086.33,
        'open': 1087.2,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 3, tzinfo=datetime.timezone.utc),
        'volume': 44463.0,
      },
      <class 'dict'> {
        'close': 1088.5089,
        'high': 1092.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 1088.5,
        'open': 1089.66,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 4, tzinfo=datetime.timezone.utc),
        'volume': 37941.0,
      },
      <class 'dict'> {
        'close': 1090.5,
        'high': 1091.84,
        'interval': TimeFrame(duration='1Min'),
        'low': 1088.68,
        'open': 1088.68,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 5, tzinfo=datetime.timezone.utc),
        'volume': 36206.0,
      },
    ],
  }
---
# name: test_get_alpaca_bars[TSLA-stock-2021-11-26T15-00-00]
  <class 'dict'> {
    '__root__': <class 'list'> [
      <class 'dict'> {
        'close': 1093.3,
        'high': 1095.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 1091.64,
        'open': 1093.62,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 0, tzinfo=datetime.timezone.utc),
        'volume': 88087.0,
      },
      <class 'dict'> {
        'close': 1090.66,
        'high': 1093.6826,
        'interval': TimeFrame(duration='1Min'),
        'low': 1088.91,
        'open': 1093.04,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 1, tzinfo=datetime.timezone.utc),
        'volume': 53702.0,
      },
      <class 'dict'> {
        'close': 1087.21,
        'high': 1090.6399,
        'interval': TimeFrame(duration='1Min'),
        'low': 1087.2,
        'open': 1090.2,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 2, tzinfo=datetime.timezone.utc),
        'volume': 50956.0,
      },
      <class 'dict'> {
        'close': 1089.9759,
        'high': 1090.49,
        'interval': TimeFrame(duration='1Min'),
        'low': 1086.33,
        'open': 1087.2,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 3, tzinfo=datetime.timezone.utc),
        'volume': 44463.0,
      },
      <class 'dict'> {
        'close': 1088.5089,
        'high': 1092.0,
        'interval': TimeFrame(duration='1Min'),
        'low': 1088.5,
        'open': 1089.66,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 4, tzinfo=datetime.timezone.utc),
        'volume': 37941.0,
      },
      <class 'dict'> {
        'close': 1090.5,
        'high': 1091.84,
        'interval': TimeFrame(duration='1Min'),
        'low': 1088.68,
        'open': 1088.68,
        'timestamp': datetime.datetime(2021, 11, 26, 15, 5, tzinfo=datetime.timezone.utc),
        'volume': 36206.0,
      },
    ],
  }
---
